app/app.go

```diff
@@ -236,26 +271,174 @@ func (a *App) HandleWork() {
 func (a *App) HandleTransaction(txn *blockchain.Transaction) {
 	validTransaction := a.Pool.Push(txn, a.Chain)
 	if validTransaction {
-		log.Debug("added transaction to pool from address: " + txn.Sender.Repr())
+		log.Debug("Added transaction to pool from address: " + txn.Sender.Repr())
 	} else {
-		log.Debug("bad transaction rejected from sender: " + txn.Sender.Repr())
+		log.Debug("Bad transaction rejected from sender: " + txn.Sender.Repr())
 	}
 }
 
 // HandleBlock handles new instance of BlockWork.
 func (a *App) HandleBlock(blk *blockchain.Block) {
+	log.Info("Received new block")
 	validBlock := a.Pool.Update(blk, a.Chain)
 
-	if validBlock {
-		// Append to the chain before requesting
-		// the next block so that the block
-		// numbers make sense.
-		a.Chain.AppendBlock(blk)
-		address := a.CurrentUser.Wallet.Public()
-		blk := a.Pool.NextBlock(a.Chain, address, a.CurrentUser.BlockSize)
-		if miner.IsMining() {
-			miner.RestartMiner(a.Chain, blk)
+	if !validBlock {
+		// The block was invalid wrt our chain. Maybe our chain is out of date.
+		// Update it and try again.
+		chainChanged, err := a.SyncBlockChain()
+		if chainChanged && miner.IsMining() {
+			miner.StopMining()
+			go a.Mine()
+		}
+		if err != nil {
+			log.WithError(err).Error("Error synchronizing blockchain")
+			return
+		}
+
+		validBlock = a.Pool.Update(blk, a.Chain)
+		if !validBlock {
+			// Synchronizing our chain didn't help, the block is still invalid.
+			return
+		}
+	}
+
+	// Append to the chain before requesting the next block so that the block
+	// numbers make sense.
+	a.Chain.AppendBlock(blk)
+	if miner.IsMining() {
+		miner.StopMining()
+		go a.Mine()
+	}
+	log.Debug("Added block number %d to chain", blk.BlockNumber)
+	return
+}
+
+// Mine continuously pulls transactions form the transaction pool, uses them to
+// create blocks, and mines those blocks. When a block is mined it is added
+// to the blockchain and broadcasted into the network. Mine() returns when
+// miner.StopMining() is called.
+func (a *App) Mine() {
+	log.Info("Starting miner")
+	for {
+		// Make a new block form the transactions in the transaction pool
+		blockToMine := a.Pool.NextBlock(a.Chain, a.CurrentUser.Wallet.Public(),
+			a.CurrentUser.BlockSize)
+		miningResult := miner.Mine(a.Chain, blockToMine)
+		if miningResult.Complete {
+			log.Info("Sucessfully mined a block!")
+			push := msg.Push{
+				ResourceType: msg.ResourceBlock,
+				Resource:     blockToMine,
+			}
+			a.PeerStore.Broadcast(push)
+		} else if miningResult.Info == miner.MiningHalted {
+			log.Info("Miner stopped.")
+			return
+		}
+	}
+}
+
+// SyncBlockChain updates the local copy of the blockchain by requesting missing
+// blocks from peers. Returns true if the blockchain changed as a result of
+// calling this function, false if it didn't and an error if we are not connected
+// to any peers.
+func (a *App) SyncBlockChain() (bool, error) {
```

Comment:
It's possible, whoever unlikely, that all our peers disconnect from us while this function is running, in which case the blockchain might have been partially updated, but the update is incomplete. Checking if we are connected to peers before calling this would not guarantee that we have peers to make requests to throughout the duration of the function. So I think we still need both return values here.